docs/docs/adrs/adr-015-partial-set-security.md

```diff
@@ -0,0 +1,293 @@
+
+So, the fields can take the following values:
+| `is_top_N` | `top_N`|
+| :--- | :---|
+| `true` | `[34, 95]` |
+| `false` | not set |
+
+
+If a proposal has those arguments wrongly set, it should get rejected in [ValidateBasic](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/types/proposal.go#L86).
+
+Note that we could easily distinguish whether a chain is _Top N_ or _Opt In_ by only having the single field `top_N` and if `top_N == 0`, then we’re in the _Opt In_ case, although for clarity, we introduce both fields.
+
+In a future version of PSS, we intend to introduce a way to modify the parameters of a consumer chain, e.g, a chain that is _Opt In_ to become _Top N—, etc.
+
+#### State & Query
+We have to augment the provider module’s state to keep track of the `is_top_N` and `top_N` values for each consumer chain. They key to store this information would be:
+
+```
+ChainIdWithLenKey(IsTopNBytePrefix, chainID)
+```
+reusing [`ChainIdWithLenKey`](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/types/keys.go#L418).
+
+Then in the [keeper](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/keeper.go) we introduce methods as follows:
+```
+func (k Keeper) SetIsTopN(ctx sdk.Context, chainID string, isTopN bool) error
+func (k Keeper) IsTopN(ctx sdk.Context, chainID string) bool
+
+// returns the N if Top N chain, otherwise an error
+func (k Keeper) GetTopN(ctx sdk.Context, chainID string) (uint8, error)
+```
+
+In a future version of PSS, we intend to extend the `gaiad query provider list-consumer-chains` query to also return information on whether a consumer chain is an Opt In or a Top N chain and with what N. This way, block explorers like [Mintscan](https://www.mintscan.io/) would show for a chain “This chain is secured by N% of Cosmos Hub.”
+
+### How do validators opt in?
+A validator can opt in by sending a new type of message that we introduce in [tx.proto](https://github.com/cosmos/interchain-security/blob/v3.3.0/proto/interchain_security/ccv/provider/v1/tx.proto#L1).
+```
+message MsgOptIn {
+// the chain id of the consumer chain to opt in to
+string chain_id = 1;
+// the provider address of the validator on Cosmos Hub
+string provider_addr = 2;
+}
+```
+Note that in a Top N consumer chain, the top N% Cosmos Hub validators are forced by default to validate the consumer chain. Nevertheless, validators in the bottom 100 - N% can opt in to validate as well.
+Additionally, note that due to undelegations, etc. a validator that is in the top N% validators might fall down to the 100 - N% in which case this validator is not forced to validate anymore. Therefore, a validator that might move out of the top N% might want to send a `MsgOptIn` to keep validating even if this happens.
```

Comment:
I agree with Jehans point. I think the more natural thing is to have people explicitly send a MsgOptOut if they were once top-N and don't want to validate anymore (they have to turn off their node anyways, since they were running it before, so I think we can also expect them to send an opt-out message when they shut it down). I think this is good because it (very mildly) incentivizes people to just keep validating(= more security for the chain)

Reply:
This could be done as well but my concern is that it will not be obvious to a validator that it has to opt out.
For example, a validator `V` belongs to the Top N but votes `No` during a `ConsumerAdditionProposal` that introduces a Top N chain. Now, assume that `V` drops below the top N% validators, `V` would be forced to send a `MsgOptOut` even though it never opted in which seems peculiar. This case in reverse already happens today, where you might have to validate even though you do not want to, but I don't see why we should introduce another variant of this.
Another (extreme) example could be the following: A validator does not want to secure a consumer chain `C` before `C` starts but the validator belongs to the Top N so decides to re-delegate half of the tokens to another validator (under its control) so that both the old and the new validator are under Top N. This validator would need to quickly send a `MsgOptOut` when the consumer chain starts so not to get jailed for downtime.

> [...] incentivizes people to just keep validating(= more security for the chain)

This was one argument in favor of the delayed opted out.

Reply:
That explanation makes sense to me, thanks for taking the time to write it out. Maybe could have some of the pros/cons of this exact choice somewhere in the ADR, or maybe it's not worth it

Reply:
Since consumer chains can't update their validator set, validators leaving the `TopN` are forced to send a `MsgOptIn` to keep validating the chain.
Maybe we should communicate to small validators the importance of manually opting in, even if they are momentarily in the TopN.

Reply:
We've decided that it should be possible to implement top-n completely in the opt in and opt out handlers. See above.

Reply:
After discussing more about this we reached to the conclusion that top N% validators are automatically opted in the case of a Top N consumer chain. A validator that belongs to the bottom `(100-N)%` would also automatically opt in if it enters (e.g., through delegations) the top N% validators. For a validator to opt out, the validator has to explicitly opt out by sending a `MsgOptOut` message.

> Depending on the reward mechanism, automatically opting a validator our may take away some of its rewards.
With the above approach we wouldn't have this issue.